blog/_src/posts/2016-05-24-measuring-gc-latencies-in-haskell-ocaml-racket.md

```diff
@@ -0,0 +1,472 @@
+1.0ms..2.2ms: ##### 17                            99.96%
+2.2ms..4.7ms:  1     (2.7ms)                     100.00%
+```
+
+As you can see, most pauses are between 220µs and 1ms, with the
+longest pause being 2.7ms.
+
+The other approach to measure latency for this program, which works on
+older OCaml versions without an instrumented runtime, is just to
+insert explicit timing calls and compute the worst-case time of an
+iteration -- as an over-approximation over the max pause time,
+assuming that the actual insertion/deletion time is small.
+
+```ocaml
+let worst = ref 0.
+let time f =
+  let before = Unix.gettimeofday () in
+  let result = f () in
+  let after = Unix.gettimeofday () in
+  worst := max !worst (after -. before);
+  result
+
+let push_msg chan high_id = time @@ fun () ->
+  let low_id = high_id - window_size in
+  let inserted = IMap.add high_id (message high_id) chan in
+  if low_id < 0 then inserted
+  else IMap.remove low_id inserted
+
+(* ..main loop.. *)
+let () = Printf.printf "Worst pause: %.2E\n" !worst
+```
+
+Running this version reports a worst-case latency of 2ms seconds on my
+machine (I use the `%E` formatter for scientific notation, so it gets
+printed as `2.03E-03`), which is in line with the instrumented
+runtime -- actually slightly lower, as the instrumentation may add
+some overhead.
+
+A downside of this poor man worst-latency computation approach is that
+we only get the worst time, not any kind of timing distribution.
+
+### Explaining OCaml results
+
+The OCaml GC has had reliable incremental phases implemented by
+default for a long time, and does not use a copying strategy for its
+old generation. It is mark&sweep, executed well, so it was
+predictable from the start that this specific benchmark would not be
+a worst-case for OCaml.
+
+The latest released OCaml version, OCaml 4.03.0, has seen work by
+Damien Doligez to improve the worst-case latency in some situations,
+motivated by the industrial use-cases of Jane Street. In particular,
+the latency *instrumentation* tools that I'm using above were
+developed by Damien on this occasion. I checked with the second
+measurement strategy that the latency is just as good on previous
+OCaml versions: this particular use-case was not in need of
+improvement before 4.03.
+
+## Racket version
+
+Max New wrote a first version of Racket port of this benchmark -- he
+had to explicitly keep track of the map count and minimum key to match
+the original GHC version. I adapted his code to my simplified variant,
+and it looks rather similar to the other implementations.
+
+```scheme
+#lang racket/base
+(require racket/match)
+
+(define window-size 200000)
+(define msg-count  2000000)
+
+(define (message n) (make-bytes 1024 (modulo n 256)))
+
+(define (push-msg chan id-high)
+  (define id-low (id-high . - . window-size))
+  (define inserted (hash-set chan id-high (message id-high)))
+  (if (id-low . < . 0) inserted
+      (hash-remove inserted id-low)))
+
+(define _
```

Comment:
Is this definition necessary? Is it just here to make it look similar to the OCaml code?